### ISSUE TITLE
Investigate tagger performance

### ISSUE THREAD

**Issue by duncanp-sonar** [SonarSource/sonarlint-visualstudio#3692]
### Description

Related to #2798. The worst effects of that ticket have been ameliorated by lazily-creating the tooltips. However, we will still be creating the same number of tags and tag spans, and the number we are creating seems very high.

In the example in the other ticket: the file has 25k lines and 3k issues. After scrolling, we had created > 100k tags.

#### Areas to look at:
* are we creating more tags than we need? e.g. there are two notes ([here](https://github.com/SonarSource/sonarlint-visualstudio/blob/c27cd936cb2ff6ebf091d74aab884e1b68ca9bf3/src/IssueViz/Editor/FilteringTaggerBase.cs#L94), and [here](https://github.com/SonarSource/sonarlint-visualstudio/blob/c27cd936cb2ff6ebf091d74aab884e1b68ca9bf3/src/IssueViz/Editor/FilteringTaggerBase.cs#L114)) in the `FilteringBaseTagger` that suggest we might be.
* streaming issue results: are we telling VS about the incremental changes, or are we telling VS everything in has changed?

Note: this behaviour does not trigger a gold bar, and scrolling behaviour in large files is acceptable (at least on our machines).

**Comment by duncanp-sonar** [comment]
### Number of tags being created
> In the example in the other ticket: the file has 25k lines and 3k issues. After scrolling, we had created > 100k tags.

There are multiple consumers of `IErrorTag`s in the editor - not just component responsible for rendering the squigglies inline.
Two other components are:

 (1) the VS Error Margin (`Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.OverviewMargin.Implementation.ErrorMarginElement`) and ,
(2) the VS file health indicator (`Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.UI.Text.Wpf.FileHealthIndicator`)

![image](https://github.com/SonarSource/sonarlint-visualstudio/assets/10757559/9e21db75-006c-4e09-bd5d-39c318752182)

Both of these components request the tags for the span of the entire file. It seems reasonable that they would have to do this as they are both visualization information about all of the issues in the file. However, that doesn't rule out the possibility that our taggers are triggering the update more frequently than required.

We are creating a large number of tags, but they are being destroyed i.e. we aren't leaking memory. See the logged output from an instrumented version of the code below:

#### Debug output
Repro:
* instrument the code using the code below
* open a file with a large number of issues
* scroll the file
```
11:45:58:995	(AAAAAA  Tag finalized. Active count: 8, total created: 6602, total disposed: 0, total finalized: 6594 
11:45:58:995	(AAAAAA  Tag finalized. Active count: 7, total created: 6602, total disposed: 0, total finalized: 6595 
11:45:58:995	(AAAAAA  Tag finalized. Active count: 6, total created: 6602, total disposed: 0, total finalized: 6596 
11:45:58:995	(AAAAAA  Tag finalized. Active count: 5, total created: 6602, total disposed: 0, total finalized: 6597 
11:45:58:995	(AAAAAA  Tag finalized. Active count: 4, total created: 6602, total disposed: 0, total finalized: 6598 
11:45:58:995	(AAAAAA  Tag finalized. Active count: 3, total created: 6602, total disposed: 0, total finalized: 6599 
``` 

#### Instrument SonarErrorTag code
```
    internal class SonarErrorTag : IErrorTag, IDisposable
    {
        internal static int totalCreated = 0;
        internal static int totalDisposed = 0;
        internal static int totalFinalized = 0;

        private readonly Lazy<object> tooltipFactory;

        public SonarErrorTag(string errorType, IAnalysisIssueBase analysisIssue,  IErrorTagTooltipProvider errorTagTooltipProvider)
        {
            if (totalCreated % 100 == 0)
            {
                GC.Collect();
            }

            System.Threading.Interlocked.Increment(ref totalCreated);
            Debug.WriteLine($"YYYYYY Tag created. Active count: {totalCreated - totalFinalized}, total created: {totalCreated}, total finalized: {totalFinalized}");

            ErrorType = errorType;
            tooltipFactory = new Lazy<object>(() => errorTagTooltipProvider.Create(analysisIssue));
        }

        public string ErrorType { get; }

        public object ToolTipContent => tooltipFactory.Value;

        public void Dispose()
        {
            // NB - the editor doesn't dispose tags, so this won't be called 
            System.Threading.Interlocked.Increment(ref totalDisposed);
            Debug.WriteLine($"(ZZZZZZ  Tag disposed. Active count: {totalCreated - totalDisposed}, total created: {totalCreated}, total disposed: {totalDisposed}");
        }

        ~SonarErrorTag()
        {
            System.Threading.Interlocked.Increment(ref totalFinalized);
            Debug.WriteLine($"(AAAAAA  Tag finalized. Active count: {totalCreated - totalFinalized}, total created: {totalCreated}, total disposed: {totalDisposed}, total finalized: {totalFinalized} ");
        }
    }
```
